### ISSUE TITLE
Better auto completion

### ISSUE THREAD

**Issue by mateusak** [godotengine/godot#16958]
Right now Godot handles auto-completion using the class the script inherits from; e.g If a script inherits `KinematicBody2D` it will auto-complete everything that belongs to it and it's ancestors (`CollisionObject2D`).

In the case it's the opposite; if your script inherits `CollisionObject2D` but is attached to a `KinematicBody2D` and you call `move_and_slide`, for example, it will still work. Same goes for variables if you use `self.desired_variable`. The auto-completion in this case does not work, but the functionality do.

TL;DR
Auto-completion should work not only with the class the script inherits, but with the class the object is attached to.

**Comment by Zylann** [comment]
Problem is, you are not supposed to access things your script doesn't inherit. It would encourage exploiting a hack which is not recommended design-wise.

**Comment by vnen** [comment]
Yeah, if you use the script in a KinematicBody2D, the script should inherit a KinematicBody2D (or a subclass of it but never a superclass). I wonder if this case should even be allowed.

I see no reason to support that kind of completion anyway. If you are doing this to reuse the script for multiple nodes that also inherit CollisionObject2D, then you couldn't use KB specific functions anyway.

**Comment by mateusak** [comment]
The point of auto-completion isn't to inform the programmers of coding guidelines. It's there to show what will work and what will not.

But I'm sure that won't suffice for you, so let me give you a real use example;

You want to make destroyable objects, so you create a `Destroyable.gd` which inherits `CollisionObject`. Then you proceed to create a monster (a `KinematicBody2D`), which, logically, can be destroyed. So it inherits `Destroyable.gd`. Since your auto-completion works, you have easy access to all it's functions normally. Now you want a destroyable beach ball (a `RigidBody2D`), and you can easily do that too using only `Destroyable.gd`

If your auto-completion does not work, then for the sake of commodity you'll have to create 2 identical scripts, one `DestroyableRigid.gd` and one `DestroyableKinematic.gd`. This is not only a bad practice, it goes against SSOT. It sucks to maintain.

This not only should be allowed, it should be enhanced, like removing the need for `self` before accessing a variable.

**Comment by Zylann** [comment]
`self` is optional.
Also, if you want your script to work both on two different node types, the problem is not about breaking OOP like you seem to want. The problem is about having those common functions in a common base.

**Comment by mateusak** [comment]
"Having those common functions in a common base" is **exactly** what OOP stands for. And it's exactly what I'm doing. "Breaking the OOP" would be create two different scripts for identical behavior.

And `self` isn't optional in this case. To access a variable from the class your script do not inherits you have to use `self`.

**Comment by Zylann** [comment]
It should be possible to use members from the class you inherit and all its bases, not its subclasses. The fact it works here is a bug due to how GDScript is doing calls on objects.

Unless I'm mistaken, OOP is not what you are doing because your script inherits CollisionObject but you want to use functions that are present ONLY in KinematicBody AND RigidBody subclasses, NOT their base, which you inherited. It's backwards inheritance. What I meant is, if you need functions from these two subclasses in particular, the solution is to have the common functions in CollisionObject or another intermediary class (PhysicsBody?), not exploit the fact GDScript is doing varcall (otherwise you could just always extend Node in every case, or not extend anything lol).

**Comment by mateusak** [comment]
Okay, other suggestion then; implement a `extends self`.
Which means it will by default extend the class of the node it is in. This way it is still true OOP and can be used on different grounds.

**Comment by ivanskodje** [comment]
Not sure if this is equivalent of this topic, but I will mention it just in case;
When you have a script that extends another script that extends a Node type, you do not get auto completion from that Node type on the script that extends the other script.

**Use Case Example:**

`extends "res://some_script.gd"

func _ready():
	pass`


`extends Node2D

func _ready():
	pass`


If you were to attempt to auto-complete the "position", you would not get it in the script that extends some_script.gd, but you would in the script that extends Node2D. :)

**Comment by vnen** [comment]
As said in the related issue (#20340), this is not the intended way to work. And since now there are checks that make this usage an error, it doesn't make sense to add it to code completion.
